logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,388 @@
+        );
+    }
+
+    public Collection<RubyIntegration.Filter> shutdownFlushers() {
+        return shutdownFlushes;
+    }
+
+    public Collection<RubyIntegration.Filter> periodicFlushers() {
+        return periodicFlushes;
+    }
+
+    public Collection<RubyIntegration.Output> outputs() {
+        return Collections.unmodifiableCollection(outputs.values());
+    }
+
+    public Collection<RubyIntegration.Filter> filters() {
+        return Collections.unmodifiableCollection(filters.values());
+    }
+
+    public Collection<IRubyObject> inputs() {
+        return inputs;
+    }
+
+    public RubyIntegration.Plugin registerPlugin(final RubyIntegration.Plugin plugin) {
+        plugin.register();
+        return plugin;
+    }
+
+    /**
+     * This method contains the actual compilation of the {@link Dataset} representing the
+     * underlying pipeline from the Queue to the outputs.
+     * @return Compiled {@link Dataset} representation of the underlying {@link PipelineIR} topology
+     */
+    public Dataset buildExecution() {
+        final Collection<Dataset> datasets = new ArrayList<>(5);
+        // We sort the leaves of the graph in a deterministic fashion before compilation.
+        // This is not strictly necessary for correctness since it will only influence the order
+        // of output events for which Logstash makes no guarantees, but it greatly simplifies
+        // testing and is no issue performance wise since compilation only happens on pipeline
+        // reload.
+        iffs.clear();
+        plugins.clear();
+        graph.getGraph().getAllLeaves().stream().filter(this::isOutput)
+            .sorted(Comparator.comparing(Vertex::hashPrefix)).forEachOrdered(
+            leaf -> datasets.add(outputDataset(leaf.getId(), flatten(Dataset.ROOT_DATASETS, leaf)))
+        );
+        return Dataset.TerminalDataset.from(datasets);
+    }
+
+    /**
+     * Sets up all Ruby outputs learnt from {@link PipelineIR}.
+     */
+    private Map<String, RubyIntegration.Output> setupOutputs() {
+        final Collection<PluginVertex> outs = graph.getOutputPluginVertices();
+        final Map<String, RubyIntegration.Output> res = new HashMap<>(outs.size());
+        outs.forEach(v -> {
+            final PluginDefinition def = v.getPluginDefinition();
+            final SourceWithMetadata source = v.getSourceWithMetadata();
+            res.put(v.getId(), pipeline.buildOutput(
+                RubyUtil.RUBY.newString(def.getName()), RubyUtil.RUBY.newFixnum(source.getLine()),
+                RubyUtil.RUBY.newFixnum(source.getColumn()), convertArgs(def)
+            ));
+        });
+        return res;
+    }
+
+    /**
+     * Sets up all Ruby filters learnt from {@link PipelineIR}.
+     */
+    private Map<String, RubyIntegration.Filter> setupFilters() {
+        final Collection<PluginVertex> plugins = graph.getFilterPluginVertices();
+        final Map<String, RubyIntegration.Filter> res = new HashMap<>(plugins.size());
```

Comment:
Shouldn't it be `new HashMap<>(plugins.size(), 1)`?

I don't quite get the point of hinting the exact size without also setting the load factor to 1, since we know the map will rehash at 0.75 without that.

Obvs. this is a micro-micro optimization and doesn't really matter, but if we're going to do it we may as well do it fully no?

I'm also fine with no size hint.